pkg/kvstore/kvstore.go

```diff
@@ -0,0 +1,180 @@
+// caller to provision the namespace.
+
+type KeyValueStore struct {
+	b Backend
+
+	kv map[string][]byte
+
+	fileName string
+
+	l sync.Mutex
+}
+
+type Backend interface {
+	// Creates path and ovewrites whatever is there
+	Upload(ctx context.Context, r io.Reader, pth string) error
+
+	Download(ctx context.Context, pth string) (io.ReadCloser, error)
+
+	Mv(ctx context.Context, fromPath, toPath string) error
+}
+
+// Create a new KeyValueStore instance using the given Backend with the given
+// file.  If the file exists on the Backend, it is restored.
+func NewKeyValueStore(op trace.Operation, store Backend, fileName string) (*KeyValueStore, error) {
+	p := &KeyValueStore{
+		b:        store,
+		kv:       make(map[string][]byte),
+		fileName: fileName,
+	}
+
+	if err := p.restore(op); err != nil {
+		return nil, err
+	}
+
+	op.Infof("KeyValueStore(%s) restored %d keys", fileName, len(p.kv))
+
+	return p, nil
+}
+
+func (p *KeyValueStore) restore(op trace.Operation) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	rc, err := p.b.Download(op, p.fileName)
+	if err != nil {
+		// We need to check for 404 vs something else here.
+		op.Errorf("KeyValueStore(%s) ignoring error: %s", p.fileName, err)
+		return nil
+	}
+	defer rc.Close()
+
+	if err = json.NewDecoder(rc).Decode(&p.kv); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Add a key to the KeyValueStore with the given value.  The Add operation is
+// write-through consistent, which means it will sync the data to the backend
+// before updating the in-memory state and returning to the caller.
+func (p *KeyValueStore) Add(op trace.Operation, key string, value []byte) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	// get the old value in case we need to roll back
+	oldvalue, ok := p.kv[key]
+
+	if ok && bytes.Compare(oldvalue, value) == 0 {
+		// NOOP
+		return nil
+	}
+
+	p.kv[key] = value
+
+	if err := p.save(op); err != nil && ok {
+		// revert if failure
+		p.kv[key] = oldvalue
+		return err
+	}
+
+	return nil
+}
+
+// Get retrieves a key from the KeyValueStore.
+func (p *KeyValueStore) Get(op trace.Operation, key string) ([]byte, error) {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	v, ok := p.kv[key]
+	if !ok {
+		return []byte{}, ErrKeyNotFound
+	}
+
+	return v, nil
+}
+
+// Delete removes a key from the KeyValueStore.
+func (p *KeyValueStore) Delete(op trace.Operation, key string) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+
+	oldvalue, ok := p.kv[key]
+	if !ok {
+		return ErrKeyNotFound
+	}
+
+	delete(p.kv, key)
+
+	if err := p.save(op); err != nil {
+		// restore the key
+		p.kv[key] = oldvalue
+		return err
+	}
+
+	return nil
+}
+
+// Save persists the KeyValueStore to the Backend.
+func (p *KeyValueStore) Save(op trace.Operation) error {
+	p.l.Lock()
+	defer p.l.Unlock()
+	return p.save(op)
+}
+
+func (p *KeyValueStore) save(op trace.Operation) error {
+	buf, err := json.Marshal(p.kv)
+	if err != nil {
+		return err
+	}
+
+	// upload to an ephemeral file
+	tmpfile := p.fileName + ".tmp"
```

Comment:
That assumes we are doing http? I thought the backend was supposed to abstract that away?